Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp

```diff
@@ -0,0 +1,370 @@
+        static constexpr std::array validNewStateConditionsForEmptyPathname { URLPatternConstructorStringParserState::Search, URLPatternConstructorStringParserState::Hash };
+        if (std::find(validStateConditionsForEmptyPathname.begin(), validStateConditionsForEmptyPathname.end(), m_state) != validStateConditionsForEmptyPathname.end()
+            && std::find(validNewStateConditionsForEmptyPathname.begin(), validNewStateConditionsForEmptyPathname.end(), newState) != validNewStateConditionsForEmptyPathname.end()
+            && m_result.pathname.isNull())
+            m_result.pathname = m_protocolMatchesSpecialSchemeFlag ? "/"_s : emptyString();
+        // Set init's search to empty if conditions are met.
+        static constexpr std::array validStateConditionsForEmptySearch { URLPatternConstructorStringParserState::Protocol, URLPatternConstructorStringParserState::Authority, URLPatternConstructorStringParserState::Username, URLPatternConstructorStringParserState::Password, URLPatternConstructorStringParserState::Hostname, URLPatternConstructorStringParserState::Port, URLPatternConstructorStringParserState::Pathname };
+        if (std::find(validStateConditionsForEmptySearch.begin(), validStateConditionsForEmptySearch.end(), m_state) != validStateConditionsForEmptySearch.end()
+            && newState == URLPatternConstructorStringParserState::Hash
+            && m_result.search.isNull())
+            m_result.search = emptyString();
+    }
+
+    m_state = newState;
+    m_tokenIndex += skip;
+    m_componentStart = m_tokenIndex;
+    m_tokenIncrement = 0;
+}
+
+void URLPatternConstructorStringParser::updateState(ScriptExecutionContext& context)
+{
+    switch (m_state) {
+    case URLPatternConstructorStringParserState::Init:
+        // Look for protocol prefix
+        if (isNonSpecialPatternChar(m_tokenIndex, ':')) {
+            rewind();
+            m_state = URLPatternConstructorStringParserState::Protocol;
+        }
+        break;
+    case URLPatternConstructorStringParserState::Protocol:
+        // Look for protocol prefix
+        if (isNonSpecialPatternChar(m_tokenIndex, ':')) {
+            auto maybeMatchesSpecialSchemeProtocol = computeProtocolMatchSpecialSchemeFlag(context);
+            if (maybeMatchesSpecialSchemeProtocol.hasException())
+                break; // return maybeMatchesSpecialSchemeProtocol.releaseException(); // FIXME: Return exceptions
+            auto nextState = URLPatternConstructorStringParserState::Pathname;
+            auto skip = 1;
+            if (isAuthoritySlashesNext()) {
+                nextState = URLPatternConstructorStringParserState::Authority;
+                skip = 3;
+            } else if (m_protocolMatchesSpecialSchemeFlag)
+                nextState = URLPatternConstructorStringParserState::Authority;
+            changeState(nextState, skip);
+        }
+        break;
+    case URLPatternConstructorStringParserState::Authority:
+        // Look for identity terminator
```

Comment:
@szewai is it cool if I leave these comments? 

Basically the reason I had these comments in the first place is because the spec has a bunch of one-liner functions that are helpful for understanding what the function does but create clutter in the code. I wanted to leave these comments to make it clear what I'm doing in this code. (And so if anyone re-visits this code, they'll understand why these functions are "unimplemented"). 

eg:
https://urlpattern.spec.whatwg.org/#is-a-protocol-suffix
https://urlpattern.spec.whatwg.org/#is-an-identity-terminator
<img width="907" alt="Screenshot 2024-12-19 at 11 59 34 AM" src="https://github.com/user-attachments/assets/326839c5-07ae-424d-b0df-842139961e94" />

Reply:
Also @cdumez  too

Reply:
Sure you could leave comments (with period as Chris mentioned).